streaming/base/dataset.py

```diff
@@ -305,6 +368,54 @@ def _download_file(self, basename: str, wait: bool = False) -> str:
         download_or_wait(remote, local, wait, self.retry, self.timeout)
         return local
 
+    def _download_shard_part(self,
+                             shard: int,
+                             partition: Partition,
+                             raw_info: FileInfo,
+                             zip_info: Optional[FileInfo] = None,
+                             compression: Optional[str] = None) -> None:
+        """Download shard data given metadata for the raw and compressed versions of it.
+
+        MDS format uses joint shards (ie, one file per shard). Other formats supported by streaming
+        use split shards (ie, shard data lives in two files per shard: the raw data itself and
+        metadata in a separate file).
+
+        Args:
+            shard (int): Shard ID.
+            partition (Partition): Our rank and worker's partition of the dataset.
+            raw_info (FileInfo): Raw file info.
+            zip_info (FileInfo, optional): Zip file info. Default: ``None``.
+            compression (str, optional): Compression algorithm used for zip_info. Default: ``None``.
+        """
+        # If the local raw file already exists, this is a no-op.
+        raw_filename = os.path.join(self.local, self.split, raw_info.basename)
+        if os.path.isfile(raw_filename):
+            return
+
+        # Determine if we are the doer (process performing the download), or the waiter (one waiting
+        # for it to complete). Multiple processes share a shard when it is split between workers.
+        wait = shard not in partition.shards_to_download
+
+        # Is compression used?
+        if zip_info:
+            # Download the compressed form if missing (or wait on download).
+            zip_filename = os.path.join(self.local, self.split, zip_info.basename)
+            if not os.path.isfile(zip_filename):
+                self._download_file(zip_info.basename, wait)
+
+            # Validate and decompress (or wait on that).
+            self._decompress_shard_part(zip_info, zip_filename, raw_filename, compression, wait)
+        else:
+            # Download the raw version.
+            self._download_file(raw_info.basename, wait)
+
+            # Doer or waiter?
+            if not wait:
```

Comment:
Question: How does `wait` value getting set to `True` or `False` inside multiple dataloader workers ? is it through lock or something ?

Reply:
the source of truth here is get_partition() in index.py, which assigns balanced-sized slices of shards to workers (partitions of work)